Malformed JetStream message-schedule restore state can panic the server

### Observed behavior

Restoring a file-backed JetStream stream can crash the server if the restored
snapshot contains a malformed message scheduling state file (`msgs/sched.db`).
The normal path for creating a file-backed stream does not appear to produce
this malformed file: scheduler state is written by the server's encoder, and a
regular server-created snapshot does not intentionally synthesize a truncated
`sched.db`. The malformed state therefore requires a crafted restore archive,
manual backup modification, or storage/transfer corruption.

The restore code extracts regular files from the snapshot tarball into a
temporary stream directory and then opens the restored stream. If the restored
stream configuration has `allow_msg_schedules` enabled, filestore recovery reads
`msgs/sched.db` and passes it to `MsgScheduling.decode`.

`MsgScheduling.decode` checks the top-level 17-byte header, but then trusts the
encoded schedule count before reading each entry. If the header says at least one
entry exists and the remaining buffer is empty or otherwise truncated, the
decoder can read the next fixed-width field from an empty slice and panic with an
index-out-of-range error instead of returning a decode error.

Representative panic:

```text
panic: runtime error: index out of range [1] with length 0

goroutine ...:
encoding/binary.littleEndian.Uint16(...)
        encoding/binary/binary.go:...
github.com/nats-io/nats-server/v2/server.(*MsgScheduling).decode(...)
        server/scheduler.go:290
github.com/nats-io/nats-server/v2/server.(*fileStore).recoverMsgSchedulingState(...)
        server/filestore.go:2262
```






### Expected behavior

Malformed or truncated message scheduling state should never panic the process.
The restore/recovery path should reject the bad `sched.db` contents, return or
log a clear decode error, remove or ignore the corrupted schedule-state file, and
fall back to the existing linear scan recovery path for message schedules.


### Server and client version

  repo: nats-io/nats-server
  commit: ca6311b2e1da1f20d68ccb68f20a352f3c7799d7
  nats-server: v2.15.0-dev
  go env GOVERSION: go1.25.10

### Host environment

  OS: macOS / Darwin
  Arch: arm64
  Go: go1.25.10 via GOTOOLCHAIN=auto
  nats-server source: nats-io/nats-server main
  commit: ca6311b2e1da1f20d68ccb68f20a352f3c7799d7
  nats-server --version: v2.15.0-dev
  Container runtime: none


### Steps to reproduce

#### Minimal unit-level reproduction:

1. Add this test in the `server` package:

```go
func TestMsgSchedulingDecodeRejectsTruncatedEntry(t *testing.T) {
	ms := newMsgScheduling(func() {})
	buf := make([]byte, headerLen)
	buf[0] = 1
	binary.LittleEndian.PutUint64(buf[1:], 1) // one schedule entry claimed
	binary.LittleEndian.PutUint64(buf[9:], 0) // high sequence stamp

	_, err := ms.decode(buf)
	require_Error(t, err)
}
```

2. Run the targeted test:

```sh
go test ./server -run TestMsgSchedulingDecodeRejectsTruncatedEntry -count=1
```

3. Actual result: the test panics inside `MsgScheduling.decode` before it can
   return an error. Expected result: the decoder returns `io.ErrUnexpectedEOF` or
   another explicit decode error.

#### End-to-end JetStream restore path:

1. Start `nats-server` with JetStream enabled.
2. Create a file-backed stream whose configuration enables message schedules
   (`allow_msg_schedules: true`; this also requires rollups to be allowed).
3. Build or modify a stream snapshot tarball so that it contains the usual
   stream files (`meta.inf`, `meta.sum`, `msgs/index.db`, message block files)
   and also contains a crafted `msgs/sched.db`.
4. Use this 17-byte `msgs/sched.db` payload:

```text
byte 0:    0x01          # scheduler encoding version
bytes 1-8: uint64(1)     # schedule count says one entry exists
bytes 9-16: uint64(0)    # high sequence stamp
```

5. Restore the crafted snapshot with the JetStream stream restore API.
6. During `Account.RestoreStream`, the snapshot is extracted and the restored
   stream is opened. `newFileStore` calls `recoverMsgSchedulingState`, which
   reads `msgs/sched.db` and calls `MsgScheduling.decode`. The server panics
   while decoding the truncated schedule entry.

No consumers are required for the reproduction.

--- a/server/filestore.go
+++ b/server/filestore.go
@@ -2177,7 +2177,9 @@ func (fs *fileStore) recoverTTLState() error {
 		ttlseq, err = fs.ttls.Decode(buf)
 		if err != nil {
 			fs.warn("Error decoding TTL state: %s", err)
+			// Remove the file, and reset collected state (if any).
 			_ = os.Remove(fn)
+			fs.ttls = thw.NewHashWheel()
 		}
 	}
 
@@ -2262,7 +2264,9 @@ func (fs *fileStore) recoverMsgSchedulingState() error {
 		schedSeq, err = fs.scheduling.decode(buf)
 		if err != nil {
 			fs.warn("Error decoding message scheduling state: %s", err)
+			// Remove the file, and reset collected state (if any).
 			_ = os.Remove(fn)
+			fs.scheduling = newMsgScheduling(fs.runMsgScheduling)
 		}
 	}
 

--- a/server/filestore_test.go
+++ b/server/filestore_test.go
@@ -10935,6 +10935,40 @@ func TestFileStoreMessageScheduleEncodeDecode(t *testing.T) {
 	}
 }
 
+func TestFileStoreMessageScheduleDecodeRejectsMalformed(t *testing.T) {
+	// Build a valid header that claims a single schedule entry.
+	header := func(count uint64) []byte {
+		b := make([]byte, headerLen)
+		b[0] = 1                                    // Magic version
+		binary.LittleEndian.PutUint64(b[1:], count) // Entry count
+		binary.LittleEndian.PutUint64(b[9:], 0)     // High sequence stamp
+		return b
+	}
+
+	for _, test := range []struct {
+		title string
+		buf   []byte
+		err   error
+	}{
+		{title: "ShortHeader", buf: make([]byte, headerLen-1), err: io.ErrShortBuffer},
+		{title: "BadVersion", buf: func() []byte { b := header(0); b[0] = 2; return b }(), err: ErrMsgScheduleInvalidVersion},
+		// Claims one entry but the buffer ends right after the header.
+		{title: "TruncatedAtSubjLen", buf: header(1), err: io.ErrUnexpectedEOF},
+		// Claims one entry, has the subject length but the subject bytes are missing.
+		{title: "TruncatedSubj", buf: append(header(1), 5, 0), err: io.ErrUnexpectedEOF},
+		// Has the subject but the timestamp/seq varints are missing.
+		{title: "TruncatedVarints", buf: append(header(1), 3, 0, 'f', 'o', 'o'), err: io.ErrUnexpectedEOF},
+		// Has the subject and timestamp varint but the seq varint is missing.
+		{title: "TruncatedSeq", buf: append(append(header(1), 3, 0, 'f', 'o', 'o'), binary.AppendVarint(nil, 12345)...), err: io.ErrUnexpectedEOF},
+	} {
+		t.Run(test.title, func(t *testing.T) {
+			ms := newMsgScheduling(func() {})
+			_, err := ms.decode(test.buf)
+			require_Error(t, err, test.err)
+		})
+	}
+}
+
 func TestFileStoreCorruptedNonOrderedSequences(t *testing.T) {
 	for _, test := range []struct {
 		title   string

--- a/server/scheduler.go
+++ b/server/scheduler.go
@@ -287,6 +287,9 @@ func (ms *MsgScheduling) decode(b []byte) (uint64, error) {
 	stamp := binary.LittleEndian.Uint64(b[9:])
 	b = b[headerLen:]
 	for i := uint64(0); i < count; i++ {
+		if len(b) < 2 {
+			return 0, io.ErrUnexpectedEOF
+		}
 		sl := int(binary.LittleEndian.Uint16(b))
 		b = b[2:]
 		if len(b) < sl {
@@ -295,11 +298,11 @@ func (ms *MsgScheduling) decode(b []byte) (uint64, error) {
 		subj := string(b[:sl])
 		b = b[sl:]
 		ts, tn := binary.Varint(b)
-		if tn < 0 {
+		if tn <= 0 {
 			return 0, io.ErrUnexpectedEOF
 		}
 		seq, vn := binary.Uvarint(b[tn:])
-		if vn < 0 {
+		if vn <= 0 {
 			return 0, io.ErrUnexpectedEOF
 		}
 		ms.init(seq, subj, ts)

--- a/server/thw/thw.go
+++ b/server/thw/thw.go
@@ -230,11 +230,11 @@ func (hw *HashWheel) Decode(b []byte) (uint64, error) {
 	b = b[headerLen:]
 	for i := uint64(0); i < count; i++ {
 		ts, tn := binary.Varint(b)
-		if tn < 0 {
+		if tn <= 0 {
 			return 0, io.ErrUnexpectedEOF
 		}
 		v, vn := binary.Uvarint(b[tn:])
-		if vn < 0 {
+		if vn <= 0 {
 			return 0, io.ErrUnexpectedEOF
 		}
 		hw.Add(v, ts)

--- a/server/thw/thw_test.go
+++ b/server/thw/thw_test.go
@@ -14,6 +14,8 @@
 package thw
 
 import (
+	"encoding/binary"
+	"io"
 	"math"
 	"testing"
 	"time"
@@ -248,6 +250,36 @@ func TestHashWheelEncodeDecode(t *testing.T) {
 			require_True(t, ok)
 			require_Equal(t, ts, nts)
 		}
+	}
+}
+
+func TestHashWheelDecodeRejectsMalformed(t *testing.T) {
+	// Build a valid header that claims the given number of entries.
+	header := func(count uint64) []byte {
+		b := make([]byte, headerLen)
+		b[0] = 1                                    // Magic version
+		binary.LittleEndian.PutUint64(b[1:], count) // Entry count
+		binary.LittleEndian.PutUint64(b[9:], 0)     // High sequence stamp
+		return b
+	}
+
+	for _, test := range []struct {
+		title string
+		buf   []byte
+		err   error
+	}{
+		{title: "ShortHeader", buf: make([]byte, headerLen-1), err: io.ErrShortBuffer},
+		{title: "BadVersion", buf: func() []byte { b := header(0); b[0] = 2; return b }(), err: ErrInvalidVersion},
+		// Claims one entry but the buffer ends right after the header.
+		{title: "TruncatedAtVarints", buf: header(1), err: io.ErrUnexpectedEOF},
+		// Has the timestamp varint but the seq varint is missing.
+		{title: "TruncatedSeq", buf: append(header(1), binary.AppendVarint(nil, 12345)...), err: io.ErrUnexpectedEOF},
+	} {
+		t.Run(test.title, func(t *testing.T) {
+			hw := NewHashWheel()
+			_, err := hw.Decode(test.buf)
+			require_Error(t, err, test.err)
+		})
 	}
 }
 

